Fix intermittent secondary index IT failures

Occasionally, builds on travis will fail due to an assertion error when performing deletes on secondary index ITs.  Here are some examples of the failures:

```
testLocalIngestAndQuerySpatialTemporal - org.locationtech.geowave.test.secondary.BasicSecondaryIndexIT (dataStoreOptions=REDIS)(org.locationtech.geowave.test.secondary.BasicSecondaryIndexIT)  Time elapsed: 11.169 sec  <<< FAILURE!
java.lang.AssertionError: Unable to delete all features in bulk delete
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.locationtech.geowave.test.basic.AbstractGeoWaveBasicVectorIT.deleteInternal(AbstractGeoWaveBasicVectorIT.java:541)
	at org.locationtech.geowave.test.basic.AbstractGeoWaveBasicVectorIT.testDeleteCQL(AbstractGeoWaveBasicVectorIT.java:444)
	at org.locationtech.geowave.test.secondary.AbstractSecondaryIndexIT.testIngestAndQuery(AbstractSecondaryIndexIT.java:73)
	at org.locationtech.geowave.test.secondary.BasicSecondaryIndexIT.testIngestAndQuery(BasicSecondaryIndexIT.java:77)
	at org.locationtech.geowave.test.secondary.BasicSecondaryIndexIT.testLocalIngestAndQuerySpatialTemporal(BasicSecondaryIndexIT.java:68)
```

```
testDistributedIngestAndQuerySpatialTemporal - org.locationtech.geowave.test.secondary.MapReduceSecondaryIndexIT (dataStoreOptions=ROCKSDB) (inputDataStoreOptions=ROCKSDB)(org.locationtech.geowave.test.secondary.MapReduceSecondaryIndexIT)  Time elapsed: 17.611 sec  <<< FAILURE!
java.lang.AssertionError: Unable to delete all features in bulk delete
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.locationtech.geowave.test.basic.AbstractGeoWaveBasicVectorIT.deleteInternal(AbstractGeoWaveBasicVectorIT.java:541)
	at org.locationtech.geowave.test.basic.AbstractGeoWaveBasicVectorIT.testDeleteCQL(AbstractGeoWaveBasicVectorIT.java:444)
	at org.locationtech.geowave.test.secondary.AbstractSecondaryIndexIT.testIngestAndQuery(AbstractSecondaryIndexIT.java:73)
	at org.locationtech.geowave.test.secondary.MapReduceSecondaryIndexIT.testIngestAndQuery(MapReduceSecondaryIndexIT.java:102)
	at org.locationtech.geowave.test.secondary.MapReduceSecondaryIndexIT.testDistributedIngestAndQuerySpatialTemporal(MapReduceSecondaryIndexIT.java:149)
```

```
testNoSecondaryIndices - org.locationtech.geowave.test.secondary.SimpleQuerySecondaryIndexIT (dataStoreOptions=REDIS)(org.locationtech.geowave.test.secondary.SimpleQuerySecondaryIndexIT)  Time elapsed: 0.238 sec  <<< FAILURE!
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at org.locationtech.geowave.test.secondary.SimpleQuerySecondaryIndexIT.testNoSecondaryIndices(SimpleQuerySecondaryIndexIT.java:116)
```

It seems to happen most frequently on the Redis ITs, but I have seen it happen on most of the other data stores at one point or another.

--- a/core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseDataStore.java
+++ b/core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseDataStore.java
@@ -20,6 +20,7 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.tuple.Pair;
@@ -93,6 +94,7 @@ import org.locationtech.geowave.core.store.util.NativeEntryIteratorWrapper;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
 import com.google.common.collect.Streams;
 
 public class BaseDataStore implements DataStore {
@@ -281,7 +283,7 @@ public class BaseDataStore implements DataStore {
     final Map<Short, Set<ByteArray>> dataIdsToDelete;
     if (DeletionMode.DELETE_WITH_DUPLICATES.equals(deleteMode)
         && (baseOptions.isSecondaryIndexing())) {
-      dataIdsToDelete = new HashMap<>();
+      dataIdsToDelete = new ConcurrentHashMap<Short, Set<ByteArray>>();
     } else {
       dataIdsToDelete = null;
     }
@@ -568,8 +570,13 @@ public class BaseDataStore implements DataStore {
                   Set<ByteArray> currentDataIdsToDelete =
                       internalDataIdsToDelete.get(row.getAdapterId());
                   if (currentDataIdsToDelete == null) {
-                    currentDataIdsToDelete = Collections.synchronizedSet(new HashSet<>());
-                    internalDataIdsToDelete.put(row.getAdapterId(), currentDataIdsToDelete);
+                    synchronized (internalDataIdsToDelete) {
+                      currentDataIdsToDelete = internalDataIdsToDelete.get(row.getAdapterId());
+                      if (currentDataIdsToDelete == null) {
+                        currentDataIdsToDelete = Sets.newConcurrentHashSet();
+                        internalDataIdsToDelete.put(row.getAdapterId(), currentDataIdsToDelete);
+                      }
+                    }
                   }
                   currentDataIdsToDelete.add(dataId);
                 }

